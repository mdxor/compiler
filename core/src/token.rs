#[cfg(test)]
use serde::Serialize;

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub struct Token<T> {
  pub span: Span,
  pub value: T,
}

#[derive(Eq, PartialEq, Debug, Clone)]
#[cfg_attr(test, derive(Serialize))]
pub struct Span {
  pub start: usize,
  pub end: usize,
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum HeadingLevel {
  H1,
  H2,
  H3,
  H4,
  H5,
  H6,
}

impl HeadingLevel {
  pub fn new(level: usize) -> Option<HeadingLevel> {
    match level {
      1 => Some(HeadingLevel::H1),
      2 => Some(HeadingLevel::H2),
      3 => Some(HeadingLevel::H3),
      4 => Some(HeadingLevel::H4),
      5 => Some(HeadingLevel::H5),
      6 => Some(HeadingLevel::H6),
      _ => None,
    }
  }

  pub fn to_str(&self) -> &str {
    match self {
      HeadingLevel::H1 => "h1",
      HeadingLevel::H2 => "h2",
      HeadingLevel::H3 => "h3",
      HeadingLevel::H4 => "h4",
      HeadingLevel::H5 => "h5",
      HeadingLevel::H6 => "h6",
    }
  }
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub struct LinkDefinition<'source> {
  pub label: &'source str,
  pub url: &'source str,
  pub title: String,
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum Align {
  Left,
  Center,
  Right,
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum BlockToken {
  Paragraph {
    raws: Vec<Span>,
  },
  JSX(JSXElement),
  ATXHeading {
    raws: Vec<Span>,
    level: HeadingLevel,
  },
  SetextHeading {
    raws: Vec<Span>,
    level: HeadingLevel,
  },
  IndentedCode(Vec<Span>),
  BlankLine,
  ThematicBreak,
  BlockQuote {
    blocks: Vec<Token<BlockToken>>,
    level: usize,
  },
  FencedCode {
    meta_span: Span,
    code_spans: Vec<Span>,
  },
  List {
    ch: u8,
    is_tight: bool,
    order_span: Span,
    blocks: Vec<Token<BlockToken>>,
  },
  ListItem {
    indent: usize,
    blocks: Vec<Token<BlockToken>>,
  },
  // LinkDefinition,
  // Table,
  // TableHead,
  // TableCell(&'source str, bool, bool),
  // TableAlignment,
  // TableAlign(Align),
  // TableRow
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum InlineToken {
  TextSegment,
  MaybeLinkStart,
  MaybeEmphasis {
    ch: u8,
    repeat: usize,
    can_open: bool,
    can_close: bool,
  },
  EmphasisStart,
  EmphasisEnd,
  LinkStart {
    url: Span,
    title: Vec<Span>,
  },
  LinkEnd,
  //
  Emphasis(Vec<Token<InlineToken>>),
  Link {
    url: Span,
    title: Vec<Span>,
    text_children: Vec<Token<InlineToken>>,
  },
  Text(Vec<Span>),
  Code(Vec<Span>),
  CodeSegment,
  SoftBreak,
  HardBreak,
  // is email
  AutoLink(bool),
  JSX(JSXElement),
}

#[derive(Eq, PartialEq, Debug)]
pub struct AST<T> {
  pub span: Span,
  pub children: Vec<T>,
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum ContainerBlock {
  BlockQuote(usize),
  List(u8),
  ListItem(usize),
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum JSToken {
  Keyword(Span),
  Punctuator(Span),
  String(Span),
  Template(Span),
  Identifier(Span),
  Number(Span),
  Text(Span),
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum JSXAttr {
  Spread(Vec<JSXExpressionSegment>),
  KeyLiteralValue {
    key: Span,
    value: Span,
  },
  KeyValue {
    key: Span,
    value: Vec<JSXExpressionSegment>,
  },
  KeyTrueValue {
    key: Span,
  },
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum JSXExpressionSegment {
  JS(Span),
  Element(JSXElement),
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum JSX {
  Element(JSXElement),
  Text(Span),
  Expression(Vec<JSXExpressionSegment>),
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub struct JSXElement {
  pub tag: String,
  pub attributes: Vec<JSXAttr>,
  pub children: Vec<JSX>,
}
