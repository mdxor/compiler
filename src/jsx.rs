use lazy_static::lazy_static;
use regex::Regex;
#[cfg(test)]
use serde::Serialize;
pub fn is_match_jsx(source: &str) -> bool {
  lazy_static! {
    static ref JSX_REGEX: Regex =
      Regex::new(r"^< *[\p{L}\p{Nl}$_][\p{L}\p{Nl}$\p{Mn}\p{Mc}\p{Nd}\p{Pc}]*(\.[\p{L}\p{Nl}$_][\p{L}\p{Nl}$\p{Mn}\p{Mc}\p{Nd}\p{Pc}]*)*").unwrap();
  }
  JSX_REGEX.is_match(source)
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
pub struct Position {
  start: usize,
  end: usize,
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
enum JSXAttrValue<'a> {
  String(&'a str),
  Expression(JSXExpression<'a>),
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
struct JSXKeyValueAttr<'a> {
  key: &'a str,
  value: JSXAttrValue<'a>,
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
enum JSXAttr<'a> {
  KeyValue(JSXKeyValueAttr<'a>),
  SpreadExpression(JSXExpression<'a>),
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
pub struct JSXNode<'a> {
  name: &'a str,
  attrs: Vec<JSXAttr<'a>>,
  children: Vec<JSXChild<'a>>,
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
enum JSXExpressionChild<'a> {
  Code(&'a str),
  Node(JSXNode<'a>),
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
struct JSXExpression<'a> {
  children: Vec<JSXExpressionChild<'a>>,
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
enum JSXChild<'a> {
  Node(JSXNode<'a>),
  Expression(JSXExpression<'a>),
  Text(&'a str),
}

pub struct JSXParser<'a> {
  source: &'a str,
  inline: bool,
  offset: usize,
  pub size: usize,
}

impl<'a> JSXParser<'a> {
  pub fn new(source: &'a str, offset: usize, inline: bool) -> Self {
    JSXParser {
      source,
      inline,
      offset,
      size: 0,
    }
  }

  pub fn parse(&mut self) -> Result<JSXNode<'a>, &'a str> {
    self.skip();
    let node = self.scan_jsx_node()?;
    Ok(node)
  }

  fn scan_jsx_node(&mut self) -> Result<JSXNode<'a>, &'a str> {
    if self.cur_source().starts_with("<") {
      self.move_by_size(1);
      let tag_name = self.scan_jsx_tag()?;
      let attrs = self.scan_jsx_attributes()?;
      if self.cur_source().starts_with("/") {
        self.move_by_size(1);
        self.skip();
        if !self.cur_source().starts_with(">") {
          return Err("");
        }
        self.move_by_size(1);

        return Ok(JSXNode {
          name: tag_name,
          attrs,
          children: vec![],
        });
      } else if !self.cur_source().starts_with(">") {
        return Err("");
      }

      self.move_by_size(1);
      let children = self.scan_jsx_children()?;
      self.scan_jsx_end_tag(tag_name)?;
      Ok(JSXNode {
        name: tag_name,
        attrs,
        children,
      })
    } else {
      Err("")
    }
  }

  fn scan_jsx_end_tag(&mut self, expected_tag_name: &str) -> Result<(), &'a str> {
    let end_tag_name = self.scan_jsx_tag()?;
    if end_tag_name != expected_tag_name {
      return Err("");
    }
    self.skip();
    if !self.cur_source().starts_with(">") {
      return Err("");
    }
    self.move_by_size(1);
    Ok(())
  }

  fn scan_jsx_tag(&mut self) -> Result<&'a str, &'a str> {
    lazy_static! {
      static ref JSX_TAG_REGEX: Regex =
        Regex::new(r"^[\p{L}\p{Nl}$_][\p{L}\p{Nl}$\p{Mn}\p{Mc}\p{Nd}\p{Pc}]*(\.[\p{L}\p{Nl}$_][\p{L}\p{Nl}$\p{Mn}\p{Mc}\p{Nd}\p{Pc}]*)*").unwrap();
    }
    if let Some(caps) = JSX_TAG_REGEX.captures(&self.source[self.size..]) {
      let size = caps.get(0).unwrap().as_str().len();
      Ok(self.move_by_size(size))
    } else {
      Ok(self.move_by_size(0))
    }
  }

  fn skip(&mut self) {
    lazy_static! {
      static ref INLINE_SKIP_REGEX: Regex = Regex::new("^ *").unwrap();
    }
    lazy_static! {
      static ref BLOCK_SKIP_REGEX: Regex = Regex::new(r"^ *(\n *)?").unwrap();
    }
    let mut regex = &*BLOCK_SKIP_REGEX;
    if self.inline {
      regex = &*INLINE_SKIP_REGEX;
    }
    if let Some(caps) = regex.captures(self.cur_source()) {
      let size = caps.get(0).unwrap().as_str().len();
      self.size += size;
    }
  }

  fn cur_source(&mut self) -> &str {
    &self.source[self.size..]
  }

  fn move_by_size(&mut self, size: usize) -> &'a str {
    let str = &self.source[self.size..self.size + size];
    self.size += size;
    str
  }

  fn scan_jsx_attributes(&mut self) -> Result<Vec<JSXAttr<'a>>, &'a str> {
    let mut attrs = vec![];
    loop {
      self.skip();
      if self.cur_source().starts_with(">") || self.cur_source().starts_with("/") {
        return Ok(attrs);
      }
      let attr = self.scan_jsx_attribute()?;
      attrs.push(attr);
    }
  }

  fn scan_jsx_attribute(&mut self) -> Result<JSXAttr<'a>, &'a str> {
    lazy_static! {
      static ref ATTR_KEY_REGEX: Regex =
        Regex::new(r"^[\p{L}\p{Nl}$_][\p{L}\p{Nl}$\p{Mn}\p{Mc}\p{Nd}\p{Pc}]*").unwrap();
    }

    // <H1 {...attrs}>
    if self.cur_source().starts_with("{") {
      let spread_expression = self.scan_jsx_attribute_spread_expression()?;
      Ok(spread_expression)
    } else if let Some(caps) = ATTR_KEY_REGEX.captures(self.cur_source()) {
      let attr_key_size = caps.get(0).unwrap().as_str().len();
      let attr_key = self.move_by_size(attr_key_size);
      self.skip();

      // <H1 title="title" />
      if self.cur_source().starts_with("=") {
        self.move_by_size(1);
        let value = self.scan_jsx_value()?;
        Ok(JSXAttr::KeyValue(JSXKeyValueAttr {
          key: attr_key,
          value,
        }))
      } else {
        // <H1 inline />
        return Ok(JSXAttr::KeyValue(JSXKeyValueAttr {
          key: attr_key,
          value: JSXAttrValue::String("true"),
        }));
      }
    } else {
      return Err("");
    }
  }

  fn scan_jsx_attribute_spread_expression(&mut self) -> Result<JSXAttr<'a>, &'a str> {
    lazy_static! {
      static ref SPREAD_EXPRESSION_REGEX: Regex = Regex::new(r"^{ *[.]{3}").unwrap();
    }
    if SPREAD_EXPRESSION_REGEX.is_match(self.cur_source()) {
      let expression = self.scan_jsx_expression()?;
      Ok(JSXAttr::SpreadExpression(expression))
    } else {
      Err("")
    }
  }

  fn scan_jsx_value(&mut self) -> Result<JSXAttrValue<'a>, &'a str> {
    self.skip();
    if self.cur_source().starts_with("{") {
      let expression = self.scan_jsx_expression()?;
      return Ok(JSXAttrValue::Expression(expression));
    }
    let mut end_char_option: Option<char> = None;
    if self.cur_source().starts_with("\"") {
      end_char_option = Some('"');
    }
    if self.cur_source().starts_with("'") {
      end_char_option = Some('\'');
    }
    if let Some(end_char) = end_char_option {
      let value = self.scan_jsx_value_string(end_char)?;
      return Ok(value);
    }
    Err("")
  }

  fn scan_jsx_value_string(&mut self, end_char: char) -> Result<JSXAttrValue<'a>, &'a str> {
    let mut size = 0;
    self.move_by_size(end_char.len_utf8());
    let mut chars = self.cur_source().chars();

    loop {
      if let Some(char) = chars.next() {
        match char {
          '\n' => {
            return Err("");
          }
          _ => {
            if char == end_char {
              break;
            } else {
              size += 1;
            }
          }
        }
      } else {
        return Err("");
      }
    }
    let string = self.move_by_size(size);
    self.move_by_size(1);
    return Ok(JSXAttrValue::String(string));
  }

  // {"}"} {'}'} {`}`}
  // {{}}
  // {{'{': '}'}}
  // {() => <div></div>}
  fn scan_jsx_expression(&mut self) -> Result<JSXExpression<'a>, &'a str> {
    let mut parentheses_num = 0;
    let mut jsx_expression = JSXExpression { children: vec![] };
    loop {
      if self.cur_source().starts_with("{") {
        parentheses_num += 1;
        self.move_by_size(1);
      } else if self.cur_source().starts_with("}") {
        self.move_by_size(1);
        if parentheses_num > 0 {
          parentheses_num -= 1;
          if parentheses_num == 0 {
            break;
          }
        } else {
          return Err("");
        }
      } else if self.cur_source().starts_with("<") {
        let node = self.scan_jsx_node()?;
        jsx_expression.children.push(JSXExpressionChild::Node(node));
      } else {
        jsx_expression
          .children
          .push(self.scan_jsx_expression_code()?);
      }
    }
    Ok(jsx_expression)
  }

  fn scan_jsx_expression_code(&mut self) -> Result<JSXExpressionChild<'a>, &'a str> {
    let mut stack = vec![];
    let mut size = 0;
    let mut escaped = false;
    let mut chars = self.cur_source().chars();
    loop {
      if let Some(char) = chars.next() {
        match char {
          '<' | '{' => {
            if stack.is_empty() {
              break;
            } else {
              size += 1;
            }
          }
          '\\' => {
            size += 1;
            if escaped {
              escaped = false;
            } else if !stack.is_empty() {
              escaped = true;
            }
          }
          '\'' | '"' | '`' => {
            size += 1;
            if !escaped {
              if stack.is_empty() {
                stack.push(char);
              } else if stack.last().unwrap() == &char {
                stack.pop();
              }
            }
          }
          _ => {
            size += char.len_utf8();
            if escaped {
              escaped = false
            }
          }
        }
      } else {
        return Err("");
      }
    }
    Ok(JSXExpressionChild::Code(self.move_by_size(size)))
  }

  fn scan_jsx_children(&mut self) -> Result<Vec<JSXChild<'a>>, &'a str> {
    let mut children = vec![];
    loop {
      let cur_source = self.cur_source();
      if cur_source.starts_with("</") {
        self.move_by_size(2);
        return Ok(children);
      } else if cur_source.starts_with("<") {
        let child = self.scan_jsx_node()?;
        children.push(JSXChild::Node(child));
      } else if cur_source.starts_with("{") {
        let expression = self.scan_jsx_expression()?;
        children.push(JSXChild::Expression(expression));
      } else {
        children.push(self.scan_jsx_text_child()?);
      }
    }
  }

  fn scan_jsx_text_child(&mut self) -> Result<JSXChild<'a>, &'a str> {
    let mut chars = self.cur_source().chars();
    let mut size = 0;
    loop {
      if let Some(char) = chars.next() {
        match char {
          '<' | '{' => break,
          _ => {
            size += char.len_utf8();
          }
        }
      } else {
        return Err("");
      }
    }
    if size > 0 {
      return Ok(JSXChild::Text(self.move_by_size(size)));
    }
    Err("")
  }
}

#[cfg(test)]
mod tests {
  use super::*;

  fn parse(source: &str) -> Result<JSXNode, &str> {
    let mut jsx_parser = JSXParser::new(source, 0, true);
    jsx_parser.parse()
  }
  #[test]
  fn test_jsx_parse() {
    let cases = vec![
      "<div></div>",
      r#"<div test="true">中文测试<div>en</div></div>"#,
      r#"<div test="true" content={() => <span>content</span>}>中文测试<div>en</div></div>"#,
      "<React.Fragment></React.Fragment>",
      "<></>",
      "<SelfClosed />",
    ];
    let mut results = vec![];
    for case in &cases {
      let result = parse(case);
      results.push(result)
    }
    insta::assert_yaml_snapshot!(results);
  }
}
