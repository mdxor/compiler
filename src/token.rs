#[derive(Debug, PartialEq)]
pub enum Token<'a> {
  Heading1,
  Heading2,
  Heading3,
  Heading4,
  Heading5,
  Heading6,
  Text(&'a str),
  InlineCode,
}

pub fn get_token(token: &str) -> Token {
  match &*token {
    "#" => Token::Heading1,
    "##" => Token::Heading2,
    "###" => Token::Heading3,
    "####" => Token::Heading4,
    "#####" => Token::Heading5,
    "######" => Token::Heading6,
    _ => Token::Text(token),
  }
}
