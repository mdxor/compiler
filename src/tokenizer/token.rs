#[cfg(test)]
use serde::Serialize;

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
pub struct TokenResult<'source> {
  pub size: usize,
  pub token: Token<'source>,
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
pub enum Token<'source> {
  ATXHeading(ATXHeading<'source>),
  ThematicBreak,
  SetextHeading(SetextHeading<'source>),
  IndentedCode(IndentedCode<'source>),
  FencedCode(FencedCode<'source>),
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
pub struct ATXHeading<'source> {
  pub level: usize,
  pub raw_inlines: &'source str,
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
pub struct SetextHeading<'source> {
  pub level: usize,
  pub raw: &'source str,
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
pub struct IndentedCode<'source> {
  pub codes: Vec<&'source str>,
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
pub struct FencedCode<'source> {
  pub code: &'source str,
  pub meta_string: &'source str,
  pub language: &'source str,
}
