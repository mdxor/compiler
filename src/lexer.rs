use crate::jsx;
use crate::token;
use lazy_static::lazy_static;
use regex::Regex;

struct Lexer<'a> {
  source: &'a str,
}

impl<'a> Lexer<'a> {
  pub fn new(source: &'a str) -> Self {
    Lexer { source }
  }

  fn skip_whitespace(&mut self) -> usize {
    lazy_static! {
      static ref WHITESPACE_REGEX: Regex = Regex::new("^ +").unwrap();
    }
    if let Some(caps) = WHITESPACE_REGEX.captures(self.source) {
      let size = caps.get(0).unwrap().as_str().len();
      self.move_by(size);
      size
    } else {
      0
    }
  }

  fn move_by(&mut self, size: usize) -> &'a str {
    let result = &self.source[..size];
    self.source = &self.source[size..];
    result
  }

  fn tokenize(&mut self) -> Result<token::AST<'a>, &'a str> {
    let mut blocks = vec![];
    loop {
      if self.source.is_empty() {
        break;
      } else {
        let token = self.scan_block_token()?;
        blocks.push(token);
      }
    }
    let ast = token::AST { blocks };
    Ok(ast)
  }

  fn scan_block_token(&mut self) -> Result<token::BlockToken<'a>, &'a str> {
    lazy_static! {
      static ref HEADING_START_REGEX: Regex = Regex::new("^(#{1,6}) ").unwrap();
    }
    if self.source.starts_with("    ") {
      self.move_by(4);
      return Ok(self.scan_single_line_code());
    } else {
      self.skip_whitespace();
      if let Some(caps) = HEADING_START_REGEX.captures(self.source) {
        let size = caps.get(1).unwrap().as_str().len();
        return self.scan_heading(size);
      } else if self.source.starts_with("```") {
        return Ok(self.scan_multiple_line_code());
      }
      return self.scan_paragraph();
    }
    Err("")
  }

  fn scan_paragraph(&mut self) -> Result<token::BlockToken<'a>, &'a str> {
    let content = self.scan_inline_blocks()?;
    Ok(token::BlockToken::Paragraph(token::Paragraph { content }))
  }

  fn scan_heading(&mut self, level: usize) -> Result<token::BlockToken<'a>, &'a str> {
    self.move_by(level + 1);
    let content = self.scan_inline_blocks()?;
    let heading = token::Heading {
      level: level as u8,
      content,
    };
    Ok(token::BlockToken::Heading(heading))
  }

  fn scan_inline_blocks(&mut self) -> Result<Vec<token::InlineBlock<'a>>, &'a str> {
    self.skip_whitespace();
    let mut inline_blocks = vec![];
    loop {
      if self.source.is_empty() || self.source.starts_with("\n") {
        if self.source.starts_with("\n") {
          self.move_by(1);
        }
        return Ok(inline_blocks);
      } else if self.source.starts_with("`") {
      } else {
        inline_blocks.push(self.scan_inline_text());
      }
    }
  }

  fn scan_inline_text(&mut self) -> token::InlineBlock<'a> {
    let mut chars = self.source.chars();
    let mut text_size = 0;
    let mut text = "";
    loop {
      if let Some(char) = chars.next() {
        match char {
          '\n' | '`' => {
            text = self.move_by(text_size);
            break;
          }
          _ => {
            text_size += char.len_utf8();
          }
        }
      } else {
        text = self.move_by(text_size);
        break;
      }
    }
    token::InlineBlock::Text(text)
  }

  fn scan_single_line_by_end_char(&mut self, end_char: char) -> &'a str {
    let mut size = 0;
    let mut chars = self.source.chars();
    loop {
      if let Some(char) = chars.next() {
        if char == end_char || char == '\n' {
          break;
        } else {
          size += char.len_utf8();
        }
      } else {
        return self.move_by(size);
      }
    }
    self.move_by(size + 1)
  }

  fn scan_single_line_code(&mut self) -> token::BlockToken<'a> {
    let code = self.scan_single_line_by_end_char('\n');
    token::BlockToken::SCode(code)
  }

  fn scan_multiple_line_code(&mut self) -> token::BlockToken<'a> {
    lazy_static! {
      static ref CODE_END_REGEX: Regex = Regex::new(r"(^ {0,3}|\n {0,3})``` *\n?").unwrap();
    }
    self.move_by(3);
    let mut language = "";
    let mut metastring = "";
    let mut code = "";

    self.skip_whitespace();
    let mut code_info = self.scan_single_line_by_end_char('\n');
    if code_info.ends_with("\n") {
      code_info = &code_info[..code_info.len() - 1];
    }
    if !code_info.is_empty() {
      if let Some(index) = code_info.find(" ") {
        language = &code_info[0..index].trim();
        metastring = &code_info[index..].trim_start();
      } else {
        language = code_info.trim();
      }
    }

    if let Some(captures) = CODE_END_REGEX.captures(self.source) {
      let end_token = captures.get(0).unwrap().as_str();
      let code_size = self.source.find(end_token).unwrap();
      code = self.move_by(code_size);
      self.move_by(end_token.len());
    } else {
      code = self.move_by(self.source.len());
    }
    token::BlockToken::MCode(token::MCode {
      code,
      language,
      metastring,
    })
  }

  fn scan_jsx(&mut self, is_inline: bool) -> Result<jsx::JSXNode<'a>, &'a str> {
    let mut jsx_parser = jsx::JSXParser::new(self.source, 0, is_inline);
    let jsx_node = jsx_parser.parse();
    self.move_by(jsx_parser.size);
    jsx_node
  }
}

#[cfg(test)]
mod tests {
  use super::*;

  fn parse(source: &str) -> Result<token::AST, &str> {
    let mut lex = Lexer::new(source);
    let ast = lex.tokenize();
    ast
  }
  #[test]
  fn test_lexer_parse() {
    let cases = vec![
      "    abc",
      "    <div></div>",
      "```\ncode\n```",
      "```jsx\nlet a = 11;\n```",
      "```jsx meta\nlet a = 11;\n```",
      "# 123",
      "###### 123",
      "#123",
      "####### 123",
      "#### 123\n```\ncode\n```",
    ];
    let mut results = vec![];
    for case in &cases {
      let result = parse(case);
      results.push(result)
    }
    insta::assert_yaml_snapshot!(results);
  }
}
