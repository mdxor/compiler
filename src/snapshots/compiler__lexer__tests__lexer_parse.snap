---
source: src/lexer.rs
expression: results

---
- Ok:
    blocks:
      - SCode: abc
- Ok:
    blocks:
      - SCode: "<div></div>"
- Ok:
    blocks:
      - MCode:
          code: code
          language: ""
          metastring: ""
- Ok:
    blocks:
      - MCode:
          code: let a = 11;
          language: jsx
          metastring: ""
- Ok:
    blocks:
      - MCode:
          code: let a = 11;
          language: jsx
          metastring: meta
- Ok:
    blocks:
      - Heading:
          level: 1
          content:
            - Text: "123"
- Ok:
    blocks:
      - Heading:
          level: 6
          content:
            - Text: "123"
- Ok:
    blocks:
      - Paragraph:
          content:
            - Text: "#123"
- Ok:
    blocks:
      - Paragraph:
          content:
            - Text: "####### 123"
- Ok:
    blocks:
      - Heading:
          level: 4
          content:
            - Text: "123"
      - MCode:
          code: code
          language: ""
          metastring: ""

